exonum/src/events/network.rs

```diff
@@ -369,47 +380,78 @@ impl Listener {
-
-                    stream.for_each(move |raw| {
-                        let event = NetworkEvent::MessageReceived(addr, raw);
-                        network_tx.clone().send(event).map_err(into_other).map(drop)
-                    })
-                })
-                .map(|_| {
-                    // Ensure that holder lives until the stream ends.
-                    let _holder = holder;
+            let connection_handler = handshake
+                .listen(sock)
+                .and_then(move |(sock, key)| {
+                    let (_, stream) = sock.split();
+                    stream
+                        .into_future()
+                        .and_then(Ok)
+                        .map_err(|e| e.0)
+                        .and_then(move |(raw, stream)| Self::parse_connect_msg(raw, stream))
+                        .and_then(move |(connect, stream)| {
+                            trace!("Received handshake message={:?}", connect);
+
+                            Self::process_incoming_messages(
+                                stream,
+                                network_tx,
+                                connect,
+                                ConnectInfo {
+                                    address: addr,
+                                    public_key: key,
+                                },
+                            )
+                        })
+                        .map(|_| {
+                            // Ensure that holder lives until the stream ends.
+                            let _holder = holder;
+                        })
                 })
                 .map_err(log_error);
+
             handle.spawn(to_box(connection_handler));
             to_box(future::ok(()))
         });
 
         Ok(Listener(to_box(server)))
     }
+
+    fn parse_connect_msg<S>(raw: Option<RawMessage>, stream: S) -> Result<(Connect, S), io::Error> {
+        match raw.map(Any::from_raw) {
+            Some(Ok(Any::Connect(msg))) => Ok((msg, stream)),
+            Some(Ok(other)) => Err(other_error(&format!(
+                "First message is not Connect, got={:?}",
+                other
+            ))),
+            Some(Err(e)) => Err(into_other(e)),
+            None => Err(other_error("Incoming socket closed")),
+        }
+    }
+
+    fn process_incoming_messages<S>(
+        stream: SplitStream<S>,
+        network_tx: mpsc::Sender<NetworkEvent>,
+        connect: Connect,
+        info: ConnectInfo,
+    ) -> impl Future<Item = (), Error = io::Error>
+    where
+        S: Stream<Item = RawMessage, Error = io::Error>,
+    {
+        let address = info.address.clone();
+        let event = NetworkEvent::PeerConnected(info, connect);
+
+        let stream = network_tx
+            .clone()
+            .send(event)
+            .map_err(into_other)
+            .and_then(move |_| Ok(stream))
+            .flatten_stream();
+
+        stream.for_each(move |raw| {
```

Comment:
Sorry for stupid question / off-topic: What happens if the `Connect` message is invalid? Are messages from the stream still processed with this `for_each`?